### PR TITLE
AT4-1149 add a flag for std::chrono::abs to allow ADL to work without issues with C++17 and beyond

### PR DESCRIPTION
**Pull request by DevLumbaa**
On updating the compiler settings in Cheetah to use C++17, an error was thrown while building and that was because of the ambiguous abs function in headerfield.cpp file that has been modified in this PR. This PR fixes that issue.

### CODE REVIEW

**Review by chrisjwilliams — CHANGES_REQUESTED** [review]
This change will not not work in general as it forces the use of the std::abs, when you may want to use an abs function specialised for the type T. (grep for abs in the src code to see some of these specilazations that this change would ignore).
You may want to look up ADL (argument depenedent lookup) as per the deleted comment to understand how this construct works

**Comment by arunnaidu123** [comment]
It seems like the compiler is confused about which std::abs to prefer (c++17 has its own std::abs for std::chrono). Prefering std::abs might be a better option.

**Comment by chrisjwilliams** [comment]
If that is as you say then what you are proposing is a bit like shooting the messenger. Instead of fixing the underlying problem (std::abs for chrono types defined twice) you hack the part of the code that alerted you to the problem in order to shut it up.
I would be inclined to solve this by loading the abs definition for std::chrono types inside defined in astrotypes only when compiled with C++std<C++17 using feature based macros as we do in panda (CompilerInfo.h).

1. Create a new pss/astrotypes/utils/CompilerInfo.h file and define a flag HAS_STD_CHRONO_ABS (using __cplusplus to detect current C++ std).
2.  include CompilerInfo.h in the location where this is defined in the astrotypes library to include/exclude the relevant abs specialization as required.

This allows library users to continue using it with C++11 and 14 as they see fit, allows for self documenting the code (why have a std::abs defined in astrotypes), and provides a single place in the code to help manage compiler idiosyncrosis

**Comment by chrisjwilliams** [comment]
btw. The fact you could even compile the code with this change is a red flag that the code coverage is lacking. i.e. for a type that does not have std::abs defined and cannot auto convert to float. This is a good opportunity to fix that up, or at the very least put a ticket for it on the backlog.

**Review by chrisjwilliams — COMMENTED** [review]
By removing this function rather than putting it behind a pre-processor guard as previously suggested you are preventing this library being used in C++11 and C++14 projects.
Changes like this should have the consent of all stakeholders in this library (I believe e-science NL and Bonn are the relevant parties). Also such a change must be reflected in the documentation which explicitly state it is C++11 compatible.

I suggest to reinstate this code with the guards and maintain the flexibility for library users unless you are intending to add further C++17 only changes to the codebase where it might be worth the effort.  Alternatively fork this repo and maintain a pss only version for your own needs where you can ignore the exigencies of maintaining portability and flexibility of a community library.

**Comment by ArisKarastergiou** [comment]
Chris is right Dev, wasn't the plan a separate branch anyway?

Aris

**Comment by DevLumbaa** [comment]
Thanks for pointing out the ADL issue in your first comment Chris and the following comments.
Arun, Mitch and I did agree on Thursday to fork this astrotypes repo to create a PSS version of it which will be C++17 compliant but we have not had a chance to present/discuss this with the rest of the team yet. Therefore(in my view), this PR is more of a reference point to enable code review and test the CI/CD pipelines for astrotypes(if we have any) to see if the changes work as intended until we have had a chance to create a PSS version of astrotypes.
Would it make sense to convert this PR to a draft PR?

**Comment by DevLumbaa** [comment]
I have added a flag around the std::chrono::abs() to keep the current state of astrotypes unaltered. All the other changes have been undone.

**Review by chrisjwilliams — COMMENTED** [review]
Yes these changes will technically solve the problem. Its not the cleanest most opaque way (for us humans) of managing compiler features  as it mixes up facts about the compiler (has a feature or not) with the solution given a specific situation about what to do about it (in this particular case we supply the missing feature, but other workarounds may be more appropriate in other situations).

A very common pattern for dealing with this in portable code is using " HAS_\<FEATURE\>" macros, although there are others. This provides a consistent nomenclature, whilst not presupposing the solution which can be left to the relevant part of code. I our case guards would be  #ifndef HAS_STD_CHRONO_ABS which as well as being functional describes exactly why we are providing a std::abs definition (i.e self documenting)

 This logic/design pattern is so endemic for portable code that, for 2020 features at least, has made it into the C++ 2020 standard itself! Check out https://en.cppreference.com/w/cpp/feature_test . 

As an aside, there is the Boost::config library which provides a collection of these macros predefined for nearly all popular compilers. I don't find it very consistent and I wouldn't recommend in this case as this library is designed to work with a  wide range of Boost libraries to aid in reusability and not restrict the end user more than necessary.

File: src/cpp/pss/astrotypes/utils/CompilerInfo.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef PSS_ASTROTYPES_UTILS_COMPILERINFO_H
+#define PSS_ASTROTYPES_UTILS_COMPILERINFO_H
+
```
Comment (chrisjwilliams) [inline src/cpp/pss/astrotypes/utils/CompilerInfo.h]: Sorry forgot to mention. Each of these definitions should have heir own guards. i.e.
#ifndef HASH_STD_CHRONO_ABS

This allows library users to adapt to other compilers that you dont directly support by predefining their own versions of these macros according to their needs.

Comment (DevLumbaa) [reply src/cpp/pss/astrotypes/utils/CompilerInfo.h]: Does this look alright @chrisjwilliams ?

**Comment by chrisjwilliams** [comment]
- looks like you need to rebase on the latest master branch as it looks like there have been changes on master
- remove empty space at the ends of lines in CompilerInfo.h (in the licence text)
- your #endifs should have a matching comment to describe what its ending e.g.  `#endif //  HAS_STD_CHRONO_ABS`
- I'm not able to confirm this builds/runs so will leave it up to the other reviewers

**Review by mitchmickaliger — COMMENTED** [review]
I'm just adding a general note here, it seems that the current master branch does not build on ubuntu 20.04 or later (and therefore this feature branch doesn't build on those platforms either), but both will build on ubuntu 18.04 and centOS 7. Given that these older platforms were the platforms on which we developed astrotypes, this is not terribly surprising. We should, however, endeavour to fix this issue in general so users can build astrotypes using c++11 on newer linux platforms.

**Pull request merged**